### PR TITLE
feat: add kakehashi/captures protocol for tree-sitter query execution

### PR DESCRIPTION
**Pull request by atusy**
## Summary

Adds `kakehashi/captures/{full, full/delta, range}` — custom LSP methods that run a server-owned tree-sitter query over a document, mirroring the `textDocument/semanticTokens` interaction model. This opens the one tree-sitter capability the `kakehashi/node/*` family left unreachable: declarative `.scm` queries.

The motivating consumer is an [nvim-treesitter-context](https://github.com/nvim-treesitter/nvim-treesitter-context)-style sticky-context feature: a client asks for the `@context` captures of a document and renders the relevant ranges in a floating window, re-requesting cheaply on every cursor move and edit.

## Design (see `docs/architecture-decisions/captures-protocol.md`)

- **Server-owned, open-ended kinds**: the query is not sent by the client. A free-form `kind` (e.g. `"context"`) resolves to `queries/<lang>/<kind>.scm` across the configured `searchPaths`, through the existing loader (`; inherits:`, tolerant per-pattern compilation). Kinds are defined by what files exist — deliberately *not* a `QueryKind` enum extension — so a future `captures/kinds` discovery method needs no protocol change, and nvim-treesitter-context's own query files work as-is.
- **semanticTokens-style triple**: `full` hands out a `resultId` per `(uri, kind)`; `full/delta` answers a matching `previousResultId` with a positional single-edit diff over the matches array (reusing the semantic-tokens prefix/suffix algorithm — safe here without the line-count guard because matches carry absolute ranges) and falls back to a full result otherwise; `range` scopes the walk via `QueryCursor::set_byte_range`.
- **Match-grouped results**: correlated captures within one pattern (`@context` / `@context.end`) stay together; each capture carries a trackable `NodeInfo` (composes with every `kakehashi/node/*` accessor) plus an inline LSP `Range` so bulk results need no per-capture follow-up.
- **Predicates evaluated server-side** via the existing `filter_captures` path, so results agree with semantic-token highlighting.
- **Null vs. error**: unresolvable document or missing `<kind>.scm` → `null`; broken query asset → `null` + server warning (config trouble, mirroring highlights); only a malformed `kind` name (fails the `[A-Za-z0-9_-]+` whitelist guarding the filesystem lookup) is a JSON-RPC `InvalidParams` error.
- **No `matchLimit`**: rejected on review (ADR Option F) — silent truncation poisons the delta lineage and hides incompleteness; `range` is the scoping tool, and semanticTokens has no limit parameter either.
- **Scope**: host layer only in v1; injection-aware capture collection is the reserved follow-up, sequenced the same way node-reference-protocol deferred injection (PR-1 → PR-4).

## History note

The branch first shipped `kakehashi/query` (client-supplied query string, one-shot) and then replaced it within the same PR after reviewing it against the real consumer: live features re-request constantly, and a string-keyed protocol gives the server nothing stable to cache or diff against. The pivot is recorded in the ADR (the old `query-execution-protocol.md` is deleted per the delete-on-supersede convention; its history is in this branch's commits).

## Test plan

- Unit: `query_exec` core (capture extraction, predicate filtering, byte-range scoping, tolerant compilation) and the pure delta diff (`matches_delta_edit` edge cases: identical/append/prepend/middle/removal/clear) + kind whitelist
- E2E (`tests/e2e_kakehashi_captures.rs`): full round-trip with a `context.scm` installed into a temp search path — grouped matches with ranges and `resultId`, empty-edits delta on unchanged doc, single positional edit after `didChange`, full fallback on unknown `previousResultId`, range filtering, missing-kind `null`, malformed-kind error
- `cargo test --lib`, clippy `-D warnings`, and the node-protocol e2e suites all green

### CODE REVIEW

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+            return Ok(Value::Null);
+        };
+        let result_id = next_result_id();
+        let response = json!({
+            "resultId": result_id,
+            "matches": c.matches,
+            "skipped": c.skipped,
+        });
+        self.captures_cache
+            .insert((c.uri, params.kind), (result_id, c.matches));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: This doesn't move: `json!` serializes interpolated expressions by reference (`serde_json`'s `json_internal!` wraps value positions in `to_value(&...)`), so `c.matches` is still owned afterwards and the cache insert takes it by value once. Evidence that it compiles: the tree builds under `clippy -D warnings` and the full e2e suite for these handlers passes (`tests/e2e_kakehashi_captures.rs`, 24 tests). Not applying.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                };
+                json!({ "resultId": result_id, "edits": edits })
+            }
+            // Unknown or stale previousResultId: full result (LSP convention —
+            // a server may always answer a delta request with a full result).
+            _ => json!({
+                "resultId": result_id,
+                "matches": c.matches,
+                "skipped": c.skipped,
+            }),
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs:199]: Same as the sibling thread on `full`: `json!` serializes by reference, so there is no second move — the code compiles under `clippy -D warnings` with all tests green. Not applying.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                        Some(json!({
+                            "name": c.name,
+                            "node": node,
+                            "range": { "start": start, "end": end },
+                        }))
+                    })
+                    .collect();
+                json!({ "patternIndex": m.pattern_index, "captures": captures })
+            })
+            .collect();
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs:357]: Good catch — fixed in 28b0f629. The mapping now uses `filter_map` at the match level and drops any match whose captures all failed coordinate conversion, mirroring `execute_query`'s "clients never see empty match envelopes" invariant.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+            &file_name,
+        ) {
+            Ok(parsed) => parsed,
+            Err(err) => {
+                log::debug!(
+                    target: "kakehashi::captures",
+                    "no {file_name} for {language_id}: {err}"
+                );
+                return Ok(None);
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs:304]: Agreed — fixed in 28b0f629. The handler now probes `QueryLoader::find_query_file` on failure: a missing kind file stays at `debug` (expected case), while a file that exists but fails to load (IO error, broken `; inherits:` chain) logs at `warn`, matching the documented "broken asset warns" semantics.

File: docs/language-features.md [hunk 5]
```diff
@@ -355,6 +355,86 @@ same embedding layer as the input:
...
+
+```typescript
+type CapturesResult = {
+  resultId: string;                  // hand back as previousResultId for a delta
+  matches: Match[];
+  skipped: {                         // patterns dropped by tolerant compilation
+    startLine: number;               // 1-indexed line in the query file
+    endLine: number;
+    reason: string;                  // Tree-sitter's compile error for that pattern
+  }[];
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/language-features.md:392]: Fixed in 28b0f629 — both `docs/language-features.md` and the ADR shape now state that `startLine`/`endLine` refer to the combined query string when `; inherits:` is used (parents concatenated first).

File: src/lsp/lsp_impl/kakehashi/captures.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: As with the refuted round-1 thread, the original compiled — `json!` serializes interpolated expressions by reference, so there was no second move (proven by `clippy -D warnings` + the full e2e suite passing). That said, restructured in c6dafc25 to make the ownership split explicit: the cache insert now clones up front and the response takes the originals.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+            // Unknown or stale previousResultId: full result (LSP convention —
+            // a server may always answer a delta request with a full result).
+            _ => json!({
+                "resultId": result_id,
+                "matches": c.matches,
+                "skipped": c.skipped,
+            }),
+        };
+        self.captures_cache.insert(key, (result_id, c.matches));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: Same compile-claim refutation as the sibling thread (`json!` serializes by reference; the tree built and tested green). Restructured in c6dafc25 anyway: per-branch explicit clones feed the response and the originals move into the cache — and per your perf note, the common edits branch clones only the cheap result id, never the matches.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        let previous = self.captures_cache.get(&key).map(|e| e.value().clone());
+
+        let result_id = next_result_id();
+        // Per-branch clones feed the response; the originals move into the
+        // cache insert below. The common edits branch clones only the cheap
+        // result id, never the matches.
+        let response = match previous {
+            Some((prev_id, prev_matches)) if prev_id == params.previous_result_id => {
+                let edits = match matches_delta_edit(&prev_matches, &c.matches) {
+                    None => json!([]),
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs:184]: Good catch — fixed in 14f1c3de. The handler now borrows the cached entry in place for the id comparison and the diff (no clone of the previous matches on any path); the read guard is the match scrutinee temporary and drops at the end of the statement, before the insert touches the same shard.

File: src/lsp/lsp_impl/text_document/did_close.rs [hunk 8]
```diff
@@ -35,6 +35,10 @@ impl Kakehashi {
         // Clean up region ID mappings for this document (lazy-node-identity-tracking)
         self.bridge.cleanup(&uri);
 
+        // Drop stored captures results for this document (captures-protocol);
+        // a reopened document starts a fresh resultId lineage.
+        self.captures_cache.retain(|(u, _kind), _| u != &uri);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/text_document/did_close.rs]: The tuple pattern compiled fine — match ergonomics (RFC 2005) lets `|(u, _kind), _|` match `&(Url, String)`, binding `u: &Url` (every gate has been green since this line landed). Switched to `|key, _| key.0 != uri` in 4b54352a anyway for unambiguous reading.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+            .compute_captures(&params.text_document.uri, &params.kind, None)
+            .await?;
+        let Some(c) = computed else {
+            return Ok(Value::Null);
+        };
+        let result_id = next_result_id();
+        // Clone what the delta cache needs up front; the originals then flow
+        // into the response, keeping the ownership split explicit.
+        self.captures_cache
+            .insert((c.uri, params.kind), (result_id.clone(), c.matches.clone()));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: Real race — fixed in 4b54352a. The insert is now guarded on the document still being in the DocumentStore, so a request in flight across didClose no longer resurrects lineage state; a close racing past the check leaves at most one entry, which the next didClose for that URI clears again.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+            }
+            // Unknown or stale previousResultId: full result (LSP convention —
+            // a server may always answer a delta request with a full result).
+            _ => json!({
+                "resultId": result_id.clone(),
+                "matches": c.matches.clone(),
+                "skipped": c.skipped,
+            }),
+        };
+        self.captures_cache.insert(key, (result_id, c.matches));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: Fixed in 4b54352a with the same guard as `full`: the delta handler skips the cache insert when the document is no longer in the DocumentStore.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+        if self.documents.get(&c.uri).is_some() {
+            self.captures_cache
+                .insert((c.uri, params.kind), (result_id.clone(), c.matches.clone()));
+        }
+        Ok(json!({
+            "resultId": result_id,
+            "matches": c.matches,
+            "skipped": c.skipped,
+        }))
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs:155]: Agreed (and noted — this confirms the by-reference `json!` semantics from the earlier move threads). Fixed in 6d3938be: the response is built first and the originals move into the cache insert, so the matches are materialized exactly once per request. Same applied to the delta fallback branch.

File: src/lsp/lsp_impl/kakehashi/captures.rs [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+            None => None,
+            Some(r) => {
+                let start = mapper.position_to_byte_clamped(r.start);
+                let end = mapper.position_to_byte_clamped(r.end);
+                if start > end {
+                    return Ok(None);
+                }
+                Some(start..end)
+            }
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/kakehashi/captures.rs]: Fixed in 6d3938be — an inverted range is now normalized to [min, max], matching range_formatting's behavior, and null stays reserved for unresolvable document/kind states.
